### ISSUE TITLE
Extend @Replaces and @Requires with a generic type argument for a bean

### ISSUE THREAD

**Issue by ivyazmitinov** [micronaut-projects/micronaut-core#7512]
### Feature description

Since micronaut 3.0 it is possible to qualify beans via generic type arguments. It is a great feature that reduces amount of boilerplate qualifiers, but it is not integrated with another DI annotations:

1. `@Replaces` has a `named` and `qualifier` members, but not the generic type, so it is impossible to replace a bean qualified by a generic type
2. `@Requires` doesn't support qualifiers of any kind, what my limit complex configurations based on presence of specifically qualified beans

Extending these annotations with a generic type/all qualifiers will improve the coherence and usability of the DI API

**Comment by SergeyLadutko** [comment]
Very good idea!!!

**Comment by graemerocher** [comment]
sounds like a reasonable request

**Comment by dstepanov** [comment]
You know you can simply replace the bean by defining the same qualifier and add the order annotation with a higher priority. The only problem is when someone is fetching all the beans of type.

**Comment by ivyazmitinov** [comment]
@dstepanov, that's a great workaround, but a `@Replaces` annotation would define intentions of a developer who performs a replacement in a much more clear and unambiguous way

**Comment by dstepanov** [comment]
Right, I’m planning to implement `@Alternative` in the way CDI spec defines it

**Comment by graemerocher** [comment]
Triage: this remains a valid DI expressiveness enhancement. `@Replaces` and `@Requires` currently do not expose generic-type qualifiers, forcing replacement/configuration intent into lower-level ordering or custom bean definitions. Suggested solution: add a stable qualifier/generic-type member with clear matching semantics, while preserving the existing named/qualifier behavior and covering replacement plus conditional presence of parameterized beans.
